### DISCUSSION TITLE
Error handling in socket.use 3.x versions

### DISCUSSION THREAD

**Discussion by epexa** [socketio/socket.io#3865] [Q&A] [answered]
Hello!

In the architecture of my project, there are public methods that do not require access, which are mixed with methods that require access.

Examples methods on client-side:
```js
socket.emit('api.version', console.log); // public

socket.emit('events.get', console.log); // private
```

In 2.x versions there was a method `socket.use` which return: `packet` and `next`.

I could check if the method is public, because in each request I could get the method name. Like this:
```js
io.on('connection', (socket) => {

	socket.use((packet, next) => {
		if (packet[0]) {
			const methodName = packet[0];
			const isPublicMethod = [
				'api.schema',
				'api.version',
			].includes(methodName);
			if ( ! isPublicMethod) return next(new Error('Need API key!'));
			else next();
		}
	});
```

How can I do this in 3.x versions?

This code is not working, I am getting an error: "**Missing error handler on `socket`.**".

In method `socket.onAny` (which return: `event` and `...args`) cannot be called `next`.

For example, there is a method `io.use` (which return: `socket` and `next`) but it is not suitable as the event is called once on connection, not every time the methods are called.

**Comment by darrachequesne** [comment, marked as answer]
Hi! I've updated the documentation here: https://socket.io/docs/v4/server-socket-instance/#Socket-middlewares

Could you please check if that's clear enough?

**Reply by epexa** [reply]
Yes, that's what I need!
Thanks a lot!
